internal/jobqueue/dbjobqueue/dbjobqueue.go

```diff
@@ -0,0 +1,390 @@
+package dbjobqueue
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+
+	"github.com/osbuild/osbuild-composer/internal/jobqueue"
+)
+
+const (
+	sqlNotify   = `NOTIFY jobs`
+	sqlListen   = `LISTEN jobs`
+	sqlUnlisten = `UNLISTEN jobs`
+
+	sqlEnqueue = `INSERT INTO jobs(id, type, args, queued_at) VALUES ($1, $2, $3, NOW())`
+	sqlDequeue = `
+		UPDATE jobs
+		SET token = $1, started_at = now()
+		WHERE id = (
+		  SELECT id
+		  FROM ready_jobs
+			  -- use ANY here, because "type in ()" doesn't work with bound parameters
+			  -- literal syntax for this is '{"a", "b"}': https://www.postgresql.org/docs/13/arrays.html
+		  WHERE type = ANY($2)
+		  LIMIT 1
+		  FOR UPDATE SKIP LOCKED
+		)
+		RETURNING id, token, type, args`
+
+	sqlInsertDependency  = `INSERT INTO job_dependencies VALUES ($1, $2)`
+	sqlQueryDependencies = `
+		SELECT dependency_id
+		FROM job_dependencies
+		WHERE job_id = $1`
+
+	sqlQueryJob = `
+		SELECT type, args, finished_at, canceled
+		FROM jobs
+		WHERE id = $1`
+	sqlQueryJobStatus = `
+		SELECT result, queued_at, started_at, finished_at, canceled
+		FROM jobs
+		WHERE id = $1`
+	sqlQueryId = `
+                SELECT id
+                FROM jobs
+                WHERE token = $1`
+	sqlFinishJob = `
+		UPDATE jobs
+		SET finished_at = now(), result = $1, token = $2
+		WHERE id = $3 AND finished_at IS NULL`
+	sqlCancelJob = `
+		UPDATE jobs
+		SET canceled = TRUE
+		WHERE id = $1 AND finished_at IS NULL`
+
+	sqlInsertHeartbeat = `
+                INSERT INTO heartbeats(token, id, heartbeat)
+                VALUES ($1, $2, now())`
+	sqlQueryHeartbeats = `
+                SELECT token
+                FROM heartbeats
+                WHERE age(now(), heartbeat) > $1`
+	sqlRefreshHeartbeat = `
+                UPDATE heartbeats
+                SET heartbeat = now()
+                WHERE token = $1`
+	sqlDeleteHeartbeat = `
+                DELETE FROM heartbeats
+                WHERE id = $1`
+)
+
+type dbJobQueue struct {
+	pool *pgxpool.Pool
+}
+
+// Create a new dbJobQueue object for `url`.
+func New(url string) (*dbJobQueue, error) {
+	pool, err := pgxpool.Connect(context.Background(), url)
+	if err != nil {
+		return nil, fmt.Errorf("error establishing connection: %v", err)
+	}
+
+	return &dbJobQueue{pool}, nil
+}
+
+func (q *dbJobQueue) Close() {
+	q.pool.Close()
+}
+
+func (q *dbJobQueue) Enqueue(jobType string, args interface{}, dependencies []uuid.UUID) (uuid.UUID, error) {
+	conn, err := q.pool.Acquire(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error connecting to database: %v", err)
+	}
+	defer conn.Release()
+
+	tx, err := conn.Begin(context.Background())
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("error starting database transaction: %v", err)
+	}
+	defer func() { _ = tx.Rollback(context.Background()) }()
```

Comment:
I would honestly prefer logging even these rollback errors. I think that `Rollback` returns an error only for a closed transaction which should never happen but just in case.